File: androidApp/src/main/java/com/medhelp/medhelp/ui/analysis_price_list/AnalysisPriceListPresenter.java
package com.medhelp.medhelp.ui.analysis_price_list;

import android.content.Context;
import com.medhelp.medhelp.data.model.analise_price.AnalisePriceResponse;
import com.medhelp.medhelp.data.network.NetworkManager;
import com.medhelp.medhelp.data.pref.PreferencesManager;
import com.medhelp.medhelp.utils.rx.AppSchedulerProvider;
import com.medhelp.medhelp.utils.rx.SchedulerProvider;
import com.medhelp.medhelp.utils.timber_log.LoggingTree;
import com.medhelp.shared.model.UserResponse;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import io.reactivex.disposables.CompositeDisposable;
import timber.log.Timber;

public class AnalysisPriceListPresenter implements AnalisePriceListHelper.Presenter {
    private SchedulerProvider schedulerProvider;

    private AnalisePriceListHelper.View activityHelper;
    private Context context;

    PreferencesManager prefManager;
    NetworkManager networkManager;

    public AnalysisPriceListPresenter(Context context, AnalisePriceListHelper.View activityHelper)
    {
        this.context=context;

        prefManager=new PreferencesManager(context);
        networkManager=new NetworkManager(prefManager);

        schedulerProvider=new AppSchedulerProvider();

        this.activityHelper=activityHelper;
    }

    @Override
    public void getAnalisePrice() {
        activityHelper.showLoading();
        CompositeDisposable cd=new CompositeDisposable();
        cd.add(networkManager.getAnalisePrice()
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe(response ->{
                    List<String> group=separateGroup(response.getResponse());
                    activityHelper.updateView(group , response.getResponse());
                    activityHelper.hideLoading();
                },throwable -> {
                    activityHelper.hideLoading();
                    Timber.tag("my").e(LoggingTree.getMessageForError(throwable,"AnalysisPriceListPresenter$getAnalisePrice "));
                    activityHelper.showErrorScreen();
                })
        );
    }

    @Override
    public void removePassword() {
        prefManager.setCurrentPassword("");
        prefManager.setUsersLogin(null);
        prefManager.setCurrentUserInfo(null);
    }

    @Override
    public String getUserToken() {
        return prefManager.getCurrentUserInfo().getApiKey();
    }


    private List<String> separateGroup(List<AnalisePriceResponse> list)
    {
        List<String> groupList=new ArrayList<>();

        groupList.add(list.get(0).getGroup());

        for (AnalisePriceResponse tmp : list)
        {
            Boolean overlap=false;

            for (String str : groupList)
            {
                if(str.equals(tmp.getGroup()))
                {
                    overlap=true;
                    break;
                }
            }

            if(!overlap)
            {
                groupList.add(tmp.getGroup());
            }

        }

        Collections.sort(groupList);
        return groupList;
    }

    @Override
    public UserResponse getCurrentUser()
    {
        return prefManager.getCurrentUserInfo();
    }

    @Override
    public void setCurrentUser(UserResponse user)
    {
        prefManager.setCurrentUserInfo(user);
    }
}

File: androidApp/src/main/java/com/medhelp/medhelp/ui/analysis_price_list/AnalysisPriceListPresenter.kt
package com.medhelp.medhelp.ui.analysis_price_list

import com.medhelp.medhelp.utils.timber_log.LoggingTree.Companion.getMessageForError
import com.medhelp.medhelp.data.pref.PreferencesManager
import com.medhelp.newmedhelp.model.AnalisePriceResponse
import io.reactivex.disposables.CompositeDisposable

import com.medhelp.shared.model.UserResponse
import com.medhelp.shared.network.NetworkManager
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import timber.log.Timber
import java.util.*

class AnalysisPriceListPresenter(val mainView: AnalysisPriceListFragment) {
    var networkManager = NetworkManager()
    var prefManager = PreferencesManager (mainView.requireContext())
    val mainScope = MainScope()

    fun getAnalisePrice() {
        mainView.showLoading()

        mainScope.launch {
            kotlin.runCatching {
                networkManager.getAnalisePrice(prefManager.currentUserInfo!!.apiKey!!, prefManager.centerInfo!!.db_name!!,prefManager.currentUserInfo!!.idUser.toString(),prefManager.currentUserInfo!!.idBranch.toString())
            }
                .onSuccess {
                    val group = separateGroup(it.response)
                    mainView.updateView(group, it.response)
                    mainView.hideLoading()
                }.onFailure {
                    mainView.hideLoading()
                    Timber.tag("my").e(getMessageForError(it, "AnalysisPriceListPresenter\$getAnalisePrice "))
                    mainView.showErrorScreen()
                }
        }
    }

    fun removePassword() {
        prefManager.currentPassword = ""
        prefManager.usersLogin = null
        prefManager.currentUserInfo = null
    }

    fun getUserToken(): String {
        return prefManager.currentUserInfo!!.apiKey!!
    }

    private fun separateGroup(list: List<AnalisePriceResponse>): List<String> {
        val groupList: MutableList<String> = ArrayList()
        groupList.add(list[0].group!!)
        for (tmp in list) {
            var overlap = false
            for (str in groupList) {
                if (str == tmp.group) {
                    overlap = true
                    break
                }
            }
            if (!overlap) {
                groupList.add(tmp.group!!)
            }
        }
        Collections.sort(groupList)
        return groupList
    }

    fun getCurrentUser(): UserResponse {
        return prefManager.currentUserInfo!!
    }

    fun setCurrentUser(user: UserResponse) {
        prefManager.currentUserInfo = user
    }
}